diff --git a/src/netlib.c b/src/netlib.c
index 206e002..8433d59 100644
--- a/src/netlib.c
+++ b/src/netlib.c
@@ -888,7 +888,7 @@ void
 stop_timer()
 {
 #ifndef WIN32
-  alarm(0);
+  //alarm(0);
 #else
   /* at some point we may need some win32 equivalent */
   if (hAlarm != (HANDLE) INVALID_HANDLE_VALUE) {
@@ -1095,6 +1095,7 @@ void
 start_timer(int time)
 {
 
+    return;
 #ifdef WIN32
   /*+*+SAF What if StartTimer is called twice without the first timer */
   /*+*+SAF expiring? */
